pkg/azure/provider.go

```diff
@@ -424,46 +428,106 @@ func (p *Provider) GetKeyVaultObjectContent(ctx context.Context, objectType stri
 		switch keybundle.Key.Kty {
 		case kv.RSA:
-			// NOTE: we are writing the RSA modulus content of the key
-			return *keybundle.Key.N, nil
+			// decode the base64 bytes for n
+			nb, err := base64.RawURLEncoding.DecodeString(*keybundle.Key.N)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			// decode the base64 bytes for e
+			eb, err := base64.RawURLEncoding.DecodeString(*keybundle.Key.E)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			e := new(big.Int).SetBytes(eb).Int64()
+			pKey := &rsa.PublicKey{
+				N: new(big.Int).SetBytes(nb),
+				E: int(e),
+			}
+			derBytes, err := x509.MarshalPKIXPublicKey(pKey)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			pubKeyBlock := &pem.Block{
+				Type:  "PUBLIC KEY",
+				Bytes: derBytes,
+			}
+			var pemData []byte
+			pemData = append(pemData, pem.EncodeToMemory(pubKeyBlock)...)
+			return string(pemData), nil
+		case kv.EC:
+			// decode the base64 bytes for x
+			xb, err := base64.RawURLEncoding.DecodeString(*keybundle.Key.X)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			// decode the base64 bytes for y
+			yb, err := base64.RawURLEncoding.DecodeString(*keybundle.Key.Y)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			crv, err := getCurve(keybundle.Key.Crv)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			pKey := &ecdsa.PublicKey{
+				X:     new(big.Int).SetBytes(xb),
+				Y:     new(big.Int).SetBytes(yb),
+				Curve: crv,
+			}
+			derBytes, err := x509.MarshalPKIXPublicKey(pKey)
+			if err != nil {
+				return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
+			}
+			pubKeyBlock := &pem.Block{
+				Type:  "PUBLIC KEY",
+				Bytes: derBytes,
+			}
+			var pemData []byte
+			pemData = append(pemData, pem.EncodeToMemory(pubKeyBlock)...)
+			return string(pemData), nil
 		default:
 			err := errors.Errorf("failed to get key. key type '%s' currently not supported", keybundle.Key.Kty)
 			return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
 		}
 	case VaultObjectTypeCertificate:
+		// for object type "cert" the certificate is written to the file in PEM format
 		certbundle, err := kvClient.GetCertificate(ctx, *vaultURL, objectName, objectVersion)
 		if err != nil {
 			return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
 		}
-		if !*certbundle.Policy.KeyProperties.Exportable {
-			err := errors.Errorf("cert key is not exportable")
-			return "", wrapObjectTypeError(err, objectType, objectName, objectVersion)
-		}
-		secretBundle, err := kvClient.GetSecret(ctx, *vaultURL, objectName, objectVersion)
```

Comment:
NOTE: The sync k8s secrets logic in the csi driver will need to be updated to look for the private key from the secret instead of the cert. This is not backward compatible. We should make a note of this in the readme and the release note after we cut a new release from both repos.

Reply:
There wouldn't be a change in logic, right? Instead of the object type being `cert`, the user would need to change object type to `secret` to get both the private key and certificate? https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/sample/ingress-controller-tls/secretproviderclass-azure-tls.yaml#L21-L22

Reply:
Opened an issue to track that - https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/106

Reply:
Yea you are right. I thought there was an e2e that tests this logic which would need to be updated, but I guess not. We should add one :)